lib2/client/okta.go

```diff
@@ -0,0 +1,667 @@
+	var oktaCreds OktaCredential
+
+	item, err := kr.Get("okta-creds")
+	if err != nil {
+		return oktaCreds, err
+	}
+
+	if err = json.Unmarshal(item.Data, &oktaCreds); err != nil {
+		return oktaCreds, fmt.Errorf("Failed to get okta credentials from your keyring.  Please make sure you have added okta credentials with `aws-okta add`")
+	}
+	return oktaCreds, nil
+}
+
+// looks up the okta domain based on the region. For example, the okta domain
+// for "us" is `okta.com` making your api domain as `<your-org>.okta.com`
+func GetOktaDomain(region string) (string, error) {
+	switch region {
+	case "us":
+		return OktaServerUs, nil
+	case "emea":
+		return OktaServerEmea, nil
+	case "preview":
+		return OktaServerPreview, nil
+	}
+	return "", fmt.Errorf("invalid region %s", region)
+}
+
+// Creates and initializes an OktaClient. This is intended to provide a simple
+// way to create a client that can make requests to the Okta APIs.
+//
+// As an example for how a client might be used:
+// This client can then be passed to a provider that will manage auth
+// for other platforms. Currently AWS SAML provider is supported to get STS
+// credentials to get access to AWS services.
+//
+// Supported configuration options:
+//       TODO: expand on configuration options and add tests.
+//
+// -- proxy config: TBD
+// -- session caching: Passing in a keyring will enable support for caching.
+//      this will cache a valid okta session securely in the keyring. This
+//			session is only for access to the Okta APIs, any additional sessions
+//			(for example, aws STS credentials) will be cached by the provider that
+//      creates them.
+func NewOktaClient(creds OktaCredential, kr *keyring.Keyring, mfaConfig MFAConfig) (*OktaClient, error) {
+
+	if creds.IsValid() {
+		log.Debug("Credentials are valid :", creds.Username, " @ ", creds.Domain)
+	} else {
+		return nil, errors.New("credentials aren't complete. To remedy this, re-add your credentials with `aws-okta add`")
+	}
+
+	// url parse & set base
+	base, err := url.Parse(fmt.Sprintf(
+		"https://%s", creds.Domain,
+	))
+	if err != nil {
+		return nil, err
+	}
+
+	jar, err := cookiejar.New(&cookiejar.Options{PublicSuffixList: publicsuffix.List})
+	if err != nil {
+		return nil, err
+	}
+
+	transCfg := &http.Transport{
+		Proxy:               http.ProxyFromEnvironment,
+		TLSHandshakeTimeout: Timeout,
+	}
+
+	client := http.Client{
+		Transport: transCfg,
+		Timeout:   Timeout,
+		Jar:       jar,
+	}
+
+	oktaClient := OktaClient{
+		creds: creds,
+		// remove this now that we're storing the client
+		CookieJar: jar,
+		BaseURL:   base,
+		MFAConfig: mfaConfig,
+		UserAuth:  &OktaUserAuthn{},
+		Keyring:   kr,
+		client:    client,
+	}
+
+	// this can fail if we don't have have a backend defined.
+	// failing to retrived a cached cookie shouldn't fail the entire
+	// operation.
+	err = oktaClient.retrieveSessionCookie()
+	if err != nil {
+		// log the error to debug and continue
```

Comment:
Oof, doubt this is desireable

Reply:
agree, handling the error and make more explanation.